File: 10.0.3.020/examples/cdftf/source/cdft_example_support.f90
!*******************************************************************************
!                              INTEL CONFIDENTIAL
!   Copyright(C) 2005-2008 Intel Corporation. All Rights Reserved.
!   The source code contained  or  described herein and all documents related to
!   the source code ("Material") are owned by Intel Corporation or its suppliers
!   or licensors.  Title to the  Material remains with  Intel Corporation or its
!   suppliers and licensors. The Material contains trade secrets and proprietary
!   and  confidential  information of  Intel or its suppliers and licensors. The
!   Material  is  protected  by  worldwide  copyright  and trade secret laws and
!   treaty  provisions. No part of the Material may be used, copied, reproduced,
!   modified, published, uploaded, posted, transmitted, distributed or disclosed
!   in any way without Intel's prior express written permission.
!   No license  under any  patent, copyright, trade secret or other intellectual
!   property right is granted to or conferred upon you by disclosure or delivery
!   of the Materials,  either expressly, by implication, inducement, estoppel or
!   otherwise.  Any  license  under  such  intellectual property  rights must be
!   express and approved by Intel in writing.
!
!*******************************************************************************
!   Content:
!       MKL Cluster DFT example support functions (Fortran-interface)
!
!*******************************************************************************

      SUBROUTINE Dfti_Example_Status_Print(STATUS)

      USE MKL_DFTI

      INTEGER STATUS
      CHARACTER(DFTI_MAX_MESSAGE_LENGTH) Error_Message
      LOGICAL Class_Error

      Class_Error = DftiErrorClass(STATUS, DFTI_ERROR_CLASS)
      IF (.NOT. Class_Error) THEN
       PRINT *,' Status is not a member of Predefined Error Class'
      ELSE
       Error_Message = DftiErrorMessage(STATUS)
       PRINT *, ' Error_Message = ', Error_Message
      ENDIF

      END SUBROUTINE

      SUBROUTINE PRINT_DATA_2D_Z(X, M, N, C)

       COMPLEX(8) X(:,:)
       INTEGER M, N, C, I, J

        DO J = 1, N
         PRINT '(/" Row ",I3,":"/" "\)', J
         DO I = 1, M
	      PRINT '("(",F8.3,",",F8.3,")"\)',REAL(X(I,J)),AIMAG(X(I,J))
	      IF ((MOD(I,C).EQ.0).AND.(I.NE.M)) PRINT '(/" "\)'
	     END DO
        END DO
        PRINT *

      END SUBROUTINE

      SUBROUTINE PRINT_DATA_2D_C(X, M, N, C)

       COMPLEX(4) X(:,:)
       INTEGER M, N, C, I, J

       DO J = 1, N
		PRINT '(/" Row ",I3,":"/" "\)', J
        DO I = 1, M
			PRINT '("(",F8.3,",",F8.3,")"\)',REAL(X(I,J)),AIMAG(X(I,J))
			IF ((MOD(I,C).EQ.0).AND.(I.NE.M)) PRINT '(/" "\)'
		END DO
       END DO
       PRINT *

      END SUBROUTINE

      INTEGER FUNCTION MKL_CDFT_DATA_S(COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS,GLOBAL,NX,START_X,LOCAL,FLAG)

       IMPLICIT NONE

       INCLUDE 'mpif.h'

! MPI related integer data should all be INTEGER(4)
       INTEGER(4), ALLOCATABLE :: COUNTS(:),DISPLS(:),BUF(:)
       INTEGER(4) I,TMP(2),REQ,STAT(MPI_STATUS_SIZE)
       INTEGER(4) COMM,ROOTRANK,NPROC,NRANK,MPI_ERR

       INTEGER ELEMENTSIZE,DIM,LENGTHS(*),NX,START_X,FLAG
       COMPLEX(4) GLOBAL(*),LOCAL(*)
       INTENT(IN) COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS,NX,START_X,FLAG

       INTEGER FD

       CALL MPI_COMM_RANK(COMM,NRANK,MPI_ERR)
       IF (MPI_ERR/=MPI_SUCCESS) GOTO 100
       IF (NRANK==ROOTRANK) THEN
        CALL MPI_COMM_SIZE(COMM,NPROC,MPI_ERR)
        IF (MPI_ERR/=MPI_SUCCESS) GOTO 100
        ALLOCATE(COUNTS(NPROC),DISPLS(NPROC),BUF(2*NPROC))
       END IF

       FD=1
       DO I=1,DIM-1
        FD=FD*LENGTHS(I)
       END DO

       TMP(1)=NX*FD*ELEMENTSIZE
       TMP(2)=(START_X-1)*FD

       CALL MPI_GATHER(TMP,2,MPI_INTEGER,BUF,2,MPI_INTEGER,ROOTRANK,COMM,MPI_ERR)
       IF (MPI_ERR/=MPI_SUCCESS) GOTO 100

       IF (NRANK==ROOTRANK) THEN
        COUNTS=BUF(1:2*NPROC-1:2)
        DISPLS=BUF(2:2*NPROC:2)
       END IF

       IF (FLAG==0) THEN

        CALL MPI_IRECV(LOCAL,TMP(1),MPI_BYTE,ROOTRANK,123,COMM,REQ,MPI_ERR)
        IF (MPI_ERR/=MPI_SUCCESS) GOTO 100

        IF (NRANK==ROOTRANK) THEN
	     DO I=0,NPROC-1
              CALL MPI_SEND(GLOBAL(DISPLS(I+1)+1),COUNTS(I+1),MPI_BYTE,I,123,COMM,MPI_ERR)
	      IF (MPI_ERR/=MPI_SUCCESS) GOTO 100
	     END DO
        END IF

        CALL MPI_WAIT(REQ,STAT,MPI_ERR)
        IF (MPI_ERR/=MPI_SUCCESS) GOTO 100
       ENDIF

       IF (FLAG==1) THEN
        CALL MPI_ISEND(LOCAL,TMP(1),MPI_BYTE,ROOTRANK,222,COMM,REQ,MPI_ERR)
        IF (MPI_ERR/=MPI_SUCCESS) GOTO 100

        IF (NRANK==ROOTRANK) THEN
         DO I=0,NPROC-1
          CALL MPI_RECV(GLOBAL(DISPLS(I+1)+1),COUNTS(I+1),MPI_BYTE,I,222,COMM,STAT,MPI_ERR)
          IF (MPI_ERR/=MPI_SUCCESS) GOTO 100
         END DO
	END IF
	CALL MPI_WAIT(REQ,STAT,MPI_ERR)
       END IF

       IF (NRANK==ROOTRANK) DEALLOCATE(COUNTS,DISPLS,BUF)

100    MKL_CDFT_DATA_S=MPI_ERR

      END FUNCTION

      INTEGER FUNCTION MKL_CDFT_DATA_D(COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS,GLOBAL,NX,START_X,LOCAL,FLAG)

       IMPLICIT NONE

       INCLUDE 'mpif.h'

       INTEGER COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS(*),NX,START_X,FLAG
       COMPLEX(8) GLOBAL(*),LOCAL(*)
       INTENT(IN) COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS,NX,START_X,FLAG

       INTEGER(4) NPROC,NRANK,MPI_ERR
       INTEGER I,FD,TMP(2),REQ,STAT(MPI_STATUS_SIZE)
       INTEGER(4), ALLOCATABLE :: COUNTS(:),DISPLS(:),BUF(:)

       CALL MPI_COMM_RANK(COMM,NRANK,MPI_ERR)
       IF (MPI_ERR/=MPI_SUCCESS) GOTO 100

       IF (NRANK==ROOTRANK) THEN
        CALL MPI_COMM_SIZE(COMM,NPROC,MPI_ERR)
        IF (MPI_ERR/=MPI_SUCCESS) GOTO 100
        ALLOCATE(COUNTS(NPROC),DISPLS(NPROC),BUF(2*NPROC))
       END IF

       FD=1
       DO I=1,DIM-1
        FD=FD*LENGTHS(I)
       END DO

       TMP(1)=NX*FD*ELEMENTSIZE
       TMP(2)=(START_X-1)*FD

       CALL MPI_GATHER(TMP,2,MPI_INTEGER,BUF,2,MPI_INTEGER,ROOTRANK,COMM,MPI_ERR)
       IF (MPI_ERR/=MPI_SUCCESS) GOTO 100

       IF (NRANK==ROOTRANK) THEN
        COUNTS=BUF(1:2*NPROC-1:2)
        DISPLS=BUF(2:2*NPROC:2)
       END IF

       IF (FLAG==0) THEN

        CALL MPI_IRECV(LOCAL,TMP(1),MPI_BYTE,ROOTRANK,123,COMM,REQ,MPI_ERR)
        IF (MPI_ERR/=MPI_SUCCESS) GOTO 100

        IF (NRANK==ROOTRANK) THEN
	     DO I=0,NPROC-1
     	      CALL MPI_SEND(GLOBAL(DISPLS(I+1)+1),COUNTS(I+1),MPI_BYTE,I,123,COMM,MPI_ERR)
	      IF (MPI_ERR/=MPI_SUCCESS) GOTO 100
	     END DO
        END IF

        CALL MPI_WAIT(REQ,STAT,MPI_ERR)
        IF (MPI_ERR/=MPI_SUCCESS) GOTO 100
       ENDIF

       IF (FLAG==1) THEN

        CALL MPI_ISEND(LOCAL,TMP(1),MPI_BYTE,ROOTRANK,222,COMM,REQ,MPI_ERR)
        IF (MPI_ERR/=MPI_SUCCESS) GOTO 100

        IF (NRANK==ROOTRANK) THEN
         DO I=0,NPROC-1
          CALL MPI_RECV(GLOBAL(DISPLS(I+1)+1),COUNTS(I+1),MPI_BYTE,I,222,COMM,STAT,MPI_ERR)
          IF (MPI_ERR/=MPI_SUCCESS) GOTO 100
         END DO
	END IF
	CALL MPI_WAIT(REQ,STAT,MPI_ERR)
        IF (MPI_ERR/=MPI_SUCCESS) GOTO 100

       END IF

       IF (NRANK==ROOTRANK) DEALLOCATE(COUNTS,DISPLS,BUF)
100    MKL_CDFT_DATA_D=MPI_ERR

      END FUNCTION

      INTEGER FUNCTION MKL_CDFT_SCATTERDATA_S(COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS,GLOBAL_IN,NX,START_X,LOCAL_IN)

       INTEGER COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS(*),NX,START_X
       COMPLEX(4) GLOBAL_IN(*),LOCAL_IN(*)

       MKL_CDFT_SCATTERDATA_S=MKL_CDFT_DATA_S(COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS,GLOBAL_IN,NX,START_X,LOCAL_IN,0)

      END FUNCTION

      INTEGER FUNCTION MKL_CDFT_SCATTERDATA_D(COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS,GLOBAL_IN,NX,START_X,LOCAL_IN)

       INTEGER COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS(*),NX,START_X
       COMPLEX(8) GLOBAL_IN(*),LOCAL_IN(*)

       MKL_CDFT_SCATTERDATA_D=MKL_CDFT_DATA_D(COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS,GLOBAL_IN,NX,START_X,LOCAL_IN,0)

      END FUNCTION

      INTEGER FUNCTION MKL_CDFT_GATHERDATA_S(COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS,GLOBAL_OUT,NX,START_X,LOCAL_OUT)

       INTEGER COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS(*),NX,START_X
       COMPLEX(4) GLOBAL_OUT(*),LOCAL_OUT(*)

       MKL_CDFT_GATHERDATA_S=MKL_CDFT_DATA_S(COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS,GLOBAL_OUT,NX,START_X,LOCAL_OUT,1)

      END FUNCTION

      INTEGER FUNCTION MKL_CDFT_GATHERDATA_D(COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS,GLOBAL_OUT,NX,START_X,LOCAL_OUT)

       INTEGER COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS(*),NX,START_X
       COMPLEX(8) GLOBAL_OUT(*),LOCAL_OUT(*)

       MKL_CDFT_GATHERDATA_D=MKL_CDFT_DATA_D(COMM,ROOTRANK,ELEMENTSIZE,DIM,LENGTHS,GLOBAL_OUT,NX,START_X,LOCAL_OUT,1)

      END FUNCTION

File: 10.0.3.020/examples/cdftc/source/cdft_example_support.c
/*******************************************************************************
!                              INTEL CONFIDENTIAL
!   Copyright(C) 2005-2008 Intel Corporation. All Rights Reserved.
!   The source code contained  or  described herein and all documents related to
!   the source code ("Material") are owned by Intel Corporation or its suppliers
!   or licensors.  Title to the  Material remains with  Intel Corporation or its
!   suppliers and licensors. The Material contains trade secrets and proprietary
!   and  confidential  information of  Intel or its suppliers and licensors. The
!   Material  is  protected  by  worldwide  copyright  and trade secret laws and
!   treaty  provisions. No part of the Material may be used, copied, reproduced,
!   modified, published, uploaded, posted, transmitted, distributed or disclosed
!   in any way without Intel's prior express written permission.
!   No license  under any  patent, copyright, trade secret or other intellectual
!   property right is granted to or conferred upon you by disclosure or delivery
!   of the Materials,  either expressly, by implication, inducement, estoppel or
!   otherwise.  Any  license  under  such  intellectual property  rights must be
!   express and approved by Intel in writing.
!
!*******************************************************************************
!   Content:
!       MKL Cluster DFT interface example program (C-interface)
!
!       Examples support function set
!
!*******************************************************************************
*/

#include <stdio.h>
#include <stdlib.h>
#include "mkl_cdft.h"

/*
/   Read data from input file routine
*/
int read_data_file_2d(int argc, char *argv[], long *m, long *n)
{
    FILE *in_file;
    char *in_file_name;
    char scan_line[100];
    int  maxline=100;

    if (argc == 1) {
	printf("\n You must specify in_file data file as 1-st parameter");
	return 1;
	}
    in_file_name = argv[1];

    if( (in_file = fopen( in_file_name, "r" )) == NULL ) {
	printf("\nERROR on OPEN '%s' with mode=%s\n", in_file_name, "r");
	return 1;
    }

    fgets(scan_line, maxline, in_file);
    fscanf(in_file,"%ld \n", m);
    fgets(scan_line, maxline, in_file);
    fscanf(in_file,"%ld \n", n);
    fclose(in_file);
    return 0;
}

/*
/   Init data routines
*/
void init_data_2d_z(void *x, long m, long n)
{
    double* p = x;
    long i;

    p[0] = 1.0;
    for(i = 1; i < m*n*2; i++) p[i] = 0.0;
}

void init_data_2d_c(void *x, long m, long n)
{
    float* p = x;
    long i;

    p[0] = 1.0;
    for(i = 1; i < m*n*2; i++) p[i] = 0.0;
}

/*
/   Print data routines
*/
void print_data_2d_z(void *x, long m, long n, long c)
{
    double* p = x;
    long i, j;

    for(i = 0; i < m; i++) {
	printf("\n Row %ld:\n ",i);
	for(j = 0; j < n; j++) {
	    printf("(%8.3f,%8.3f)", p[2*j], p[2*j+1]);
	    if ((j%c == c-1) && (j != n-1)) printf("\n ");
	}
	p += 2*n;
    }
    printf("\n");
}

void print_data_2d_c(void *x, long m, long n, long c)
{
    float* p = x;
    long i, j;

    for(i = 0; i < m; i++) {
	printf("\n Row %ld:\n ",i);
	for(j = 0; j < n; j++) {
	    printf("(%8.3f,%8.3f)", p[2*j], p[2*j+1]);
	    if ((j%c == c-1) && (j != n-1)) printf("\n ");
	}
	p += 2*n;
    }
    printf("\n");
}


/*
/   Print status routine
*/
void dfti_example_status_print(long status)
{
    long   class_error;
    char*  error_message;

    class_error = DftiErrorClass(status, DFTI_ERROR_CLASS);
    if (! class_error) {
	printf(" Error Status is not a member of Predefined Error Class\n");
    } else {
	error_message = DftiErrorMessage(status);
	printf(" Error_Message = %s \n", error_message);
    }
    return;
}

/*
/   Check error routines
*/
double check_result_z(void* x, void* res_exp, long n)
{
    double* x_in = x;
    double* x_exp  = res_exp;
    double maxerr, d;
    long i;

    maxerr = 0.0;
    for (i = 0; i < 2*n; i++){
	d = x_exp[i] - x_in[i];
	if (d < 0.0) d = -d;
	if (d > maxerr) maxerr = d;
    }
    return maxerr;
}

float check_result_c(void* x, void* res_exp, long n)
{
    float* x_in = x;
    float* x_exp  = res_exp;
    float maxerr, d;
    long i;

    maxerr = 0.0;
    for (i = 0; i < 2*n; i++){
	d = x_exp[i] - x_in[i];
	if (d < 0.0) d = -d;
	if (d > maxerr) maxerr = d;
    }
    return maxerr;
}

int MKL_CDFT_Data(MPI_Comm Comm,int RootRank,int ElementSize,long Dim,long Lengths[],void *global,long nx,long start_x,void *local,int Flag) {

	int nProc,nRank,MPI_err,i,fd;
	int tmp[2],*counts,*displs,*buf;
	MPI_Request req;
	MPI_Status status;

	MPI_err=MPI_Comm_rank(Comm,&nRank);
	if (MPI_err!=MPI_SUCCESS) return MPI_err;

	if (nRank==RootRank) {
		MPI_err=MPI_Comm_size(Comm,&nProc);
		if (MPI_err!=MPI_SUCCESS) return MPI_err;
		counts=(int*)malloc(nProc*sizeof(int));
		displs=(int*)malloc(nProc*sizeof(int));
		buf=(int*)malloc(2*nProc*sizeof(int));
	}

	fd=ElementSize;
	for(i=1;i<(int)Dim;i++) fd*=(int)Lengths[i];

	tmp[0]=(int)nx*fd;
	tmp[1]=(int)start_x*fd;

	MPI_err=MPI_Gather(tmp,2,MPI_INT,buf,2,MPI_INT,RootRank,Comm);
	if (MPI_err!=MPI_SUCCESS) return MPI_err;

	if (nRank==RootRank)
	{
		for(i=0;i<nProc;i++) {
			counts[i]=buf[2*i];
			displs[i]=buf[2*i+1];
		}
	}

	if (Flag==0)
	{
		MPI_err = MPI_Irecv(local,tmp[0],MPI_BYTE, RootRank, 123,Comm,&req);
		if (MPI_err!=MPI_SUCCESS) return MPI_err;

		if (nRank==RootRank)
			for(i=0;i<nProc;i++)
			{
				MPI_err = MPI_Send(((char*)global)+displs[i],counts[i],MPI_BYTE,i,123,Comm);
				if (MPI_err!=MPI_SUCCESS) return MPI_err;
			}
		MPI_Wait(&req,&status);
		if (MPI_err!=MPI_SUCCESS) return MPI_err;
	}
	else
	{
	    MPI_err = MPI_Gatherv(local,tmp[0],MPI_BYTE,global,counts,displs,MPI_BYTE,RootRank,Comm);
	    if (MPI_err!=MPI_SUCCESS) return MPI_err;
	}

	if (nRank==RootRank) {
		free(buf);
		free(displs);
		free(counts);
	}
	return 0;
}

int MKL_CDFT_ScatterData(MPI_Comm Comm,int RootRank,int ElementSize,long Dim,long Lengths[],void *global_in,long nx,long start_x,void *local_in) {
return MKL_CDFT_Data(Comm,RootRank,ElementSize,Dim,Lengths,global_in,nx,start_x,local_in,0); }

int MKL_CDFT_GatherData(MPI_Comm Comm,int RootRank,int ElementSize,long Dim,long Lengths[],void *global_out,long nx,long start_x,void *local_out) {
return MKL_CDFT_Data(Comm,RootRank,ElementSize,Dim,Lengths,global_out,nx,start_x,local_out,1); }
